akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala

```diff
@@ -334,3 +339,85 @@ private[akka] final case class Expand[In, Out, Seed](seed: In ⇒ Seed, extrapol
   final override def restart(): Expand[In, Out, Seed] =
     throw new UnsupportedOperationException("Expand doesn't support restart")
 }
+
+/**
+ * INTERNAL API
+ */
+private[akka] final case class MapAsync[In, Out](parallelism: Int, f: In ⇒ Future[Out], decider: Supervision.Decider)
```

Comment:
I thought long and hard about this, but was unable to come up with an improvement over the status quo. You are welcome to try your hand on it after it has been merged.

A few background thoughts:

  * queue becoming full is trivially modeled by `holdUpstream`
  * queue becoming non-full requires different actions based on whether at the same time an element was dequeued or not
  * unification of the two complex tailrec expressions is impossible due to type constraints

In general layering another abstraction over the “raw” one always included a large amount of extraneous complexity and boilerplate code.